photon-ml/src/main/scala/com/linkedin/photon/ml/data/AvroDataReader.scala

```diff
@@ -0,0 +1,307 @@
+      data.drop(destColumn)
+        .withColumnRenamed(tempColName, destColumn)
+    }
+  }
+
+  /**
+   * Transforms the feature column consisting of name, term, and value feature tuples into a column of feature vectors
+   *
+   * @param df the source DataFrame
+   * @param columnName the feature column to convert into a vector
+   * @param indexMapLoader the index map loader for this column
+   * @return the transformed DataFrame
+   */
+  protected def vectorize(
+      df: DataFrame,
+      columnName: String,
+      indexMapLoader: IndexMapLoader): DataFrame = {
+
+    val vectorized = df.mapPartitions { rows =>
+      val featureMap = indexMapLoader.indexMapForRDD
+
+      rows.map { row: Row =>
+        val columnIndex = row.fieldIndex(columnName)
+        val features = row.getSeq[Row](columnIndex).flatMap {
+          case Row(featureKey: String, value: Double) =>
+            if (featureMap.contains(featureKey)) {
+              Some(featureMap.getIndex(featureKey) -> value)
+            } else {
+              None
+            }
+        }
+
+        // Add intercept if it exists in the feature map
+        val featuresWithIntercept = if (featureMap.contains(AvroDataReader.intercept._1)) {
+          features :+ (featureMap.getIndex(AvroDataReader.intercept._1) -> AvroDataReader.intercept._2)
+        } else {
+          features
+        }
+
+        val (indices, values) = featuresWithIntercept.distinct.sorted.unzip
+        val vector = new SparseVector(featureMap.featureDimension, indices.toArray, values.toArray)
+
+        // Add the column
+        Row.fromSeq(row.toSeq :+ vector)
+      }
+    }
+
+    // Update the schema for the new column
+    val vectorizedColName = tempColumnName(columnName)
+    val schema = StructType(df.schema.fields :+
+      StructField(vectorizedColName, new VectorUDT()))
+
+    sqlContext.createDataFrame(vectorized, schema)
+      .drop(columnName)
+      .withColumnRenamed(vectorizedColName, columnName)
+  }
+
+  /**
+   * Converts the feature columns consisting of name, term, and value tuples into feature vectors
+   *
+   * @param df the source DataFrame
+   * @param indexMapLoader the index map loaders for all the columns to convert
+   * @return the transformed DataFrame
+   */
+  protected def vectorize(df: DataFrame, indexMapLoaders: Map[String, IndexMapLoader]): DataFrame =
+    indexMapLoaders.keys.foldLeft(df) { (acc, columnName) =>
+      vectorize(acc, columnName, indexMapLoaders(columnName))
+    }
+
+  /**
+   * Generates a default index map loader by scanning the dataframe
+   *
+   * @param df the source DataFrame
+   * @param columnName the feature column from which to build the index
+   * @return the index map loader
+   */
+  protected def generateIndexMapLoader(df: DataFrame, columnName: String): IndexMapLoader = {
+    val features = df.select(columnName).flatMap { row: Row =>
+      row.getSeq[Row](0).map {
```

Comment:
Yes, these columns contain only one value -- the seq of row structs that represent the (key, value) pairs for features.